### ISSUE TITLE
sub shift is defined on Positionals, however Positionals need not implement `method shift`

### ISSUE THREAD

**Issue by JJ** [rakudo/rakudo#4254]
Here's the definition for `sub shift`

https://github.com/rakudo/rakudo/blob/1cd8765e6949b34a664e41a5aa717087a69da426/src/core.c/array_operators.pm6#L50

Although all core classes that implement the role do seem to have `shift`, that need not be the case for an user-defined `Positional`

## Suggestion

Although this is probably going to make it slower, re-implement it only in terms of `elems`, `AT-POS` and `EXIST-POS`. Or else, create a multi that calls one or the other depending on the actual type.

**Comment by lizmat** [comment]
I don't think you *can* implement shift in terms of `AT-POS` and `EXISTS-POS`.  That would only be possible if we also would have an `EJECT-POS` method.

**Comment by lizmat** [comment]
Ah, looking at [Array::Agnostic](https://modules.raku.org/dist/Array::Agnostic:cpan:ELIZABETH/lib/Array/Agnostic.pm6) I guess you can.  The real hard way, especially on large arrays.  Assuming that `DELETE-POS` *will* actually shorten the array when applied on the last element.

**Comment by vrurg** [comment]
By looking into `array_operators.pm6` I could say that we have a big discrepancy in signatures. Not sure if it's justified somehow. Why `pop`, `shift`, and `splice` use positionals and the rest are using raw parameters.

Anyway, I'd rather leave the current implementation of all the array ops intact and add dying method stubs on `Positional`, similarly to how it's done in `List`.

Other way around would be to have `MutablePositional` role.

**Comment by JJ** [comment]
Another option would be to create a `Shifty` role  make the argument `Shifty` and make those classes that actually implement shift mix in that role.

**Comment by vrurg** [comment]
@JJ The point I don't like `Shifty` or alike is because it binds the meaning of role to the particular class of operations. Whereas `push` and `pop`, for example, are not "shifty" in no sense. What does make these cases common is that `shift` and co. are only valid for mutable positionals.

**Comment by JJ** [comment]
El lun, 22 mar 2021 a las 16:23, Vadim Belman ***@***.***>)
escribió:

> @JJ <https://github.com/JJ> The point I don't like Shifty or alike is
> because it binds the meaning of role to the particular class of operations.
> Whereas push and pop, for example, are not "shifty" in no sense. What
> does make these cases common is that shift and co. are only valid for
> mutable positionals.
>

Wouldn't be such big a deal. it mainly defines interface, as in "Shifty has
a .shift operation defined". Then "pushy", or "poppy", all of them defined
in terms of the operations they need to implement. Or, since all three
operations seems to go together, maybe something like Queue or similar
could be a better description. My point is creating a Role and making this
a multi would not incur in a lot of overhead.

**Comment by vrurg** [comment]
I'm still curios about this case and got a bit of time to play around with `Positional`. The results are rather discouraging. 

First of all, after thoroughly considering implementation of `shift` and `pop` with friends, I came to the conclusion that we can't make them require anything but `Positional` or otherwise we gonna break a lot of 3rd party code.

With the above in mind I decided to see if I can move declarations of the required methods into `Positional`. But it also turned out to be pretty much problematic due to how typed arrays are implemented. 

A typed array is a parametrization of plain `Array` which is done by means of mixing in `Array::Typed[::TValue]` role. The latter `does Positional[TValue]`. What it means is that any methods `Positional` declares are getting overridden on the class it is [re-]applied to. The current implementation of `push`/`pop`/`shift`/etc. provided by `Array` class are totally agnostic of the array type. Type checking is handled at a lower level. Therefore, `Array::Typed` doesn't need own versions of those. But as soon as we add them to `Positional` – we loose what `Array` provides.

So, it all ends up with the only option where `Positional` is as empty as it is now.

Aside of this, I've slightly reconsidered the outcome of the following example:

```raku
class Foo is Positional { }; 
my $f = Foo.new; 
shift $f; # No such method 'shift' for invocant of type 'Foo'.  Did you mean 'unshift'?
```

So far, as we have prefix form of method invocation, the only really confusing thing is the suggestion of `unshift` method. Otherwise the error message is pretty much legit. As a matter of fact, `Foo` doesn't have the method!

More confusing it gets if one tries `unshift`:

```raku
unshift $f, 1; # Cannot resolve caller unshift(Foo:D: Int:D); none of these signatures match:    
               #    (Any:U \SELF: |values)
```

But at this particular moment I'm a bit reluctant to thinking about it.
